src/litdata/streaming/raw_dataset.py

```diff
@@ -0,0 +1,365 @@
+from litdata.utilities.dataset_utilities import generate_md5_hash, get_default_cache_dir
+
+if not _ASYNCIO_AVAILABLE:
+    raise ModuleNotFoundError(
+        "The 'asyncio' package is required for streaming datasets. Please install it with `pip install asyncio`."
+    )
+else:
+    import asyncio
+
+logger = logging.getLogger(__name__)
+SUPPORTED_PROVIDERS = ("s3", "gs", "azure")
+
+
+@dataclass
+class FileMetadata:
+    """Metadata for a single file in the dataset."""
+
+    path: str
+    size: int
+
+    def to_dict(self) -> dict[str, Any]:
+        return {"path": self.path, "size": self.size}
+
+    @classmethod
+    def from_dict(cls, data: dict[str, Any]) -> "FileMetadata":
+        return cls(path=data["path"], size=data["size"])
+
+
+class BaseIndexer(ABC):
+    """Abstract base class for file indexing strategies."""
+
+    @abstractmethod
+    def discover_files(self, input_dir: str, storage_options: Optional[dict[str, Any]]) -> list[FileMetadata]:
+        """Discover dataset files and return their metadata."""
+
+    def build_or_load_index(
+        self, input_dir: str, cache_dir: str, storage_options: Optional[dict[str, Any]]
+    ) -> list[FileMetadata]:
+        """Build or load a ZSTD-compressed index of file metadata."""
+        if not _ZSTD_AVAILABLE:
+            raise ModuleNotFoundError(str(_ZSTD_AVAILABLE))
+
+        import zstd
+
+        index_path = Path(cache_dir) / "index.json.zstd"
+
+        # Try loading cached index if it exists
+        if index_path.exists():
+            try:
+                with open(index_path, "rb") as f:
+                    compressed_data = f.read()
+                metadata = json.loads(zstd.decompress(compressed_data).decode("utf-8"))
+
+                return [FileMetadata.from_dict(file_data) for file_data in metadata["files"]]
+            except (FileNotFoundError, json.JSONDecodeError, zstd.ZstdError, KeyError) as e:
+                logger.warning(f"Failed to load cached index from {index_path}: {e}")
+
+        # Build fresh index
+        logger.info(f"Building index for {input_dir} at {index_path}")
+        files = self.discover_files(input_dir, storage_options)
+        if not files:
+            raise ValueError(f"No files found in {input_dir}")
+
+        # Cache the index with ZSTD compression
+        # TODO: upload the index to cloud storage
+        try:
+            metadata = {
+                "source": input_dir,
+                "files": [file.to_dict() for file in files],
+                "created_at": time.time(),
+            }
+            with open(os.path.join(index_path), "wb") as f:
+                f.write(zstd.compress(json.dumps(metadata).encode("utf-8")))
+        except (OSError, zstd.ZstdError) as e:
+            logger.warning(f"Error caching index to {index_path}: {e}")
+
+        logger.info(f"Built index with {len(files)} files from {input_dir} at {index_path}")
+        return files
+
+
+class FileIndexer(BaseIndexer):
```

Comment:
Let's move the file indexer to its own folder.

Reply:
Sure, keeping this for the follow-up and have added it to the list above in the description.